packages/compass-editor/src/codemirror/query-history-autocompleter.ts

```diff
@@ -21,11 +21,89 @@ export type SavedQuery = {
   };
 };
 
-export const createQueryHistoryAutocompleter = (
-  savedQueries: SavedQuery[],
-  onApply: (query: SavedQuery['queryProperties']) => void,
-  theme: CodemirrorThemeType
-): CompletionSource => {
+function simplifyQueryStringForAutocomplete(queryString: string): string {
+  return queryString.replace(/[\s'"\n\t{}},]/g, '').toLowerCase();
+}
+
+/**
+ * Codemirror runs a fuzzy search on the completion item labels.
+ * Oftentimes the fuzzy search will match on too many query history items.
+ * We limit the possible results to be improve accuracy.
+ * We give suggestions of queries that either match at least one field,
+ * or that contain the prefix the user is typing.
+ */
+function getMatchingQueryHistoryItemsForInput({
+  savedQueries,
+  queryProperty,
+  input: _input,
+}: {
+  savedQueries: SavedQuery[];
+  queryProperty: string;
+  input: string;
+}) {
+  const input = simplifyQueryStringForAutocomplete(_input);
+
+  if (input.length === 0) {
+    // Everything matches when empty search.
+    return savedQueries;
+  }
+
+  return savedQueries.filter((query) => {
+    const queryValue = query.queryProperties[queryProperty];
+
+    // Only some query properties are objects. For instance limit can be an array.
+    if (typeof queryValue !== 'object') {
+      const queryValueString = toJSString(queryValue) || '';
+      if (!queryValueString) {
+        return false;
+      }
+
+      const queryValueSimplified =
+        simplifyQueryStringForAutocomplete(queryValueString);
+      return queryValueSimplified.startsWith(input);
+    }
+
+    // We subtract parts of the string until there's nothing left.
+    // We know it's a possible match if the all of the input is found in the query.
+    let inputToMatch = input;
+
+    // We check each top level field as they can be in any order.
```

Comment:
If we're introducing a performance bottleneck for rendering with this logic, we should consider making autocompleter async and moving this code out of the main thread to prevent blocking operations on main that will have a negative effect on user interactions with the editor. Let's open a ticket for that if that's the case